File: backend/app/simulator/motion.py
import math
from datetime import datetime

from app.core.config import Settings
from app.telemetry.models import TelemetryMessage


def clamp(value, minimum, maximum):
    return max(minimum, min(maximum, value))


def move(previous: TelemetryMessage, timestamp: datetime, config: Settings):
    dt = clamp((timestamp - previous.timestamp).total_seconds(), 0, config.max_step_seconds)
    epoch = timestamp.timestamp()
    bounds = config.simulation_bounds
    phase = sum(ord(char) for char in previous.uavId)
    turn = math.sin(epoch / 20 + phase) * 1.5
    margin = min(config.boundary_margin, (bounds.east - bounds.west) / 4,
                 (bounds.north - bounds.south) / 4)
    if (previous.longitude < bounds.west + margin or previous.longitude > bounds.east - margin
            or previous.latitude < bounds.south + margin or previous.latitude > bounds.north - margin):
        target = math.degrees(math.atan2(
            ((bounds.east + bounds.west) / 2 - previous.longitude) * math.cos(math.radians(previous.latitude)),
            (bounds.north + bounds.south) / 2 - previous.latitude,
        ))
        turn = clamp((target - previous.heading + 540) % 360 - 180, -config.turn_rate, config.turn_rate)
    heading = (previous.heading + turn * dt) % 360
    distance = previous.speed * dt
    latitude = previous.latitude + distance * math.cos(math.radians(heading)) / 111195
    longitude = previous.longitude + distance * math.sin(math.radians(heading)) / (111195 * math.cos(math.radians(previous.latitude)))
    if longitude < bounds.west or longitude > bounds.east:
        heading = (360 - heading) % 360
    if latitude < bounds.south or latitude > bounds.north:
        heading = (180 - heading) % 360
    battery = max(0, previous.battery - config.battery_drain_rate * dt)
    return TelemetryMessage(**{
        **previous.model_dump(), "timestamp": timestamp,
        "longitude": clamp(longitude, bounds.west, bounds.east),
        "latitude": clamp(latitude, bounds.south, bounds.north),
        "heading": heading, "battery": battery,
        "altitude": clamp(previous.altitude + math.sin(epoch / 15 + phase) * 0.2 * dt, 20, 200),
        "signal": clamp(previous.signal + math.sin(epoch / 10 + phase) * 0.1 * dt, 0, 100),
        "status": "warning" if battery < 10 else "online",
    })

File: src/views/uav/simulator/motion.ts
import type { Uav } from "../types";
import type { UavTelemetry } from "./types";
import { simulatorConfig as config } from "./config";

const clamp = (n: number, min: number, max: number) => Math.max(min, Math.min(max, n));
const radians = Math.PI / 180;
export function canSimulate(uav: Uav) {
  const b = config.simulationBounds;
  return (
    ["online", "mission", "warning"].includes(uav.status) &&
    uav.position.longitude >= b.west &&
    uav.position.longitude <= b.east &&
    uav.position.latitude >= b.south &&
    uav.position.latitude <= b.north
  );
}
export function seedTelemetry(uav: Uav, timestamp: number): UavTelemetry {
  return {
    uavId: uav.id,
    timestamp,
    ...uav.position,
    altitude: uav.position.altitude ?? 80,
    speed: clamp(uav.telemetry.speed ?? 5, 3, 12),
    heading: uav.telemetry.heading ?? 0,
    battery: uav.telemetry.battery ?? 100,
    signal: uav.telemetry.signal ?? 95,
    status: "online",
  };
}
export function moveTelemetry(previous: UavTelemetry, timestamp: number): UavTelemetry {
  const dt = clamp((timestamp - previous.timestamp) / 1000, 0, config.maxStepSeconds);
  const b = config.simulationBounds;
  const phase = [...previous.uavId].reduce((sum, char) => sum + char.charCodeAt(0), 0);
  let turn = Math.sin(timestamp / 20000 + phase) * 1.5;
  if (
    previous.longitude < b.west + config.boundaryMargin ||
    previous.longitude > b.east - config.boundaryMargin ||
    previous.latitude < b.south + config.boundaryMargin ||
    previous.latitude > b.north - config.boundaryMargin
  ) {
    const target =
      Math.atan2(
        ((b.east + b.west) / 2 - previous.longitude) * Math.cos(previous.latitude * radians),
        (b.north + b.south) / 2 - previous.latitude
      ) / radians;
    turn = clamp(
      ((target - previous.heading + 540) % 360) - 180,
      -config.turnRate,
      config.turnRate
    );
  }
  let heading = (previous.heading + turn * dt + 360) % 360;
  const distance = previous.speed * dt;
  const latitude = previous.latitude + (distance * Math.cos(heading * radians)) / 111195;
  const longitude =
    previous.longitude +
    (distance * Math.sin(heading * radians)) / (111195 * Math.cos(previous.latitude * radians));
  if (longitude < b.west || longitude > b.east) heading = (360 - heading) % 360;
  if (latitude < b.south || latitude > b.north) heading = (180 - heading + 360) % 360;
  const battery = Math.max(0, previous.battery - config.batteryDrainRate * dt);
  return {
    ...previous,
    timestamp,
    heading,
    battery,
    longitude: clamp(longitude, b.west, b.east),
    latitude: clamp(latitude, b.south, b.north),
    altitude: clamp(previous.altitude + Math.sin(timestamp / 15000 + phase) * 0.2 * dt, 20, 200),
    signal:
      previous.signal === null
        ? null
        : clamp(previous.signal + Math.sin(timestamp / 10000 + phase) * 0.1 * dt, 0, 100),
    status: battery < 10 ? "warning" : "online",
  };
}
